browser/ui/webui/brave_rewards_internals_ui.cc

```diff
@@ -0,0 +1,128 @@
+              kBraveRewardsInternalsSettingsGeneratedSize,
+#endif
+              IDR_BRAVE_REWARDS_INTERNALS_HTML),
+      profile_(Profile::FromWebUI(web_ui)) {
+  rewards_service_ =
+      brave_rewards::RewardsServiceFactory::GetForProfile(profile_);
+  rewards_service_->AddObserver(this);
+
+  rewards_service_->GetRewardsInternalsInfo(
+      base::Bind(&BraveRewardsInternalsUI::OnGetRewardsInternalsInfo,
+                 base::Unretained(this)));
+}
+
+BraveRewardsInternalsUI::~BraveRewardsInternalsUI() {
+  rewards_service_->RemoveObserver(this);
+}
+
+void BraveRewardsInternalsUI::CustomizeWebUIProperties(
+    content::RenderViewHost* render_view_host) {
+  DCHECK(IsSafeToSetWebUIProperties());
+
+  if (render_view_host) {
+    render_view_host->SetWebUIProperty("isRewardsEnabled",
+                                       BooleanToString(IsRewardsEnabled()));
+    if (internals_info_) {
+      render_view_host->SetWebUIProperty("walletPaymentId",
+                                         internals_info_->payment_id);
+      render_view_host->SetWebUIProperty(
+          "isKeyInfoSeedValid",
+          BooleanToString(internals_info_->is_key_info_seed_valid));
+
+      base::ListValue current_reconciles;
+      for (const auto& item : internals_info_->current_reconciles) {
+        auto current_reconcile_info = std::make_unique<base::DictionaryValue>();
+        current_reconcile_info->SetString("viewingId", item.second.viewing_id_);
+        current_reconcile_info->SetString("amount", item.second.amount_);
+        current_reconcile_info->SetInteger("retryStep",
+                                           item.second.retry_step_);
+        current_reconcile_info->SetInteger("retryLevel",
+                                           item.second.retry_level_);
+        current_reconciles.Append(std::move(current_reconcile_info));
+      }
+
+      std::string json;
+      base::JSONWriter::Write(current_reconciles, &json);
+      render_view_host->SetWebUIProperty("currentReconciles", json);
+    }
+  }
+}
+
+bool BraveRewardsInternalsUI::IsRewardsEnabled() const {
+  DCHECK(profile_);
+  return profile_->GetPrefs()->GetBoolean(
+      brave_rewards::prefs::kBraveRewardsEnabled);
+}
+
+void BraveRewardsInternalsUI::UpdateWebUIProperties() {
+  if (IsSafeToSetWebUIProperties()) {
+    CustomizeWebUIProperties(GetRenderViewHost());
+    web_ui()->CallJavascriptFunctionUnsafe(
+        "brave_rewards_internals.stateUpdated");
+  }
+}
+
+void BraveRewardsInternalsUI::OnRewardsMainEnabled(
+    brave_rewards::RewardsService* rewards_service,
+    bool rewards_main_enabled) {
+  UpdateWebUIProperties();
+}
+
+void BraveRewardsInternalsUI::OnWalletInitialized(
+    brave_rewards::RewardsService* rewards_service,
+    int error_code) {
+  DCHECK(rewards_service_);
+  rewards_service_->GetRewardsInternalsInfo(
+      base::Bind(&BraveRewardsInternalsUI::OnGetRewardsInternalsInfo,
+                 base::Unretained(this)));
```

Comment:
this isn't safe because rewards service can outlive the webui page. You need WeakPtr here

Reply:
Thanks for catching.